pkg/api/helpers.go

```diff
@@ -510,35 +510,86 @@ func GetTaintsFromNodeAnnotations(annotations map[string]string) ([]Taint, error
-func TolerationToleratesTaint(toleration *Toleration, taint *Taint) bool {
-	if len(toleration.Effect) != 0 && toleration.Effect != taint.Effect {
+// ToleratesTaint checks if the toleration tolerates the taint.
+func (t *Toleration) ToleratesTaint(taint *Taint) bool {
+	if len(t.Effect) > 0 && t.Effect != taint.Effect {
 		return false
 	}
 
-	if toleration.Key != taint.Key {
+	// empty toleration key means match all taint keys
+	if len(t.Key) > 0 && t.Key != taint.Key {
 		return false
 	}
+
+	// nil ForgivenessSeconds means tolerate the taint forever
+	if t.ForgivenessSeconds != nil {
+		// taint with no added time indicated can only be tolerated
+		// by toleration with no forgivenessSeconds.
+		if taint.TimeAdded.IsZero() {
+			return false
+		}
+
+		// TODO: need to take time skew into consideration, make sure we don't evict pods earlier than we promised.
+		if unversioned.Now().After(taint.TimeAdded.Add(time.Second * time.Duration(*t.ForgivenessSeconds))) {
+			return false
+		}
+	}
+
 	// TODO: Use proper defaulting when Toleration becomes a field of PodSpec
-	if (len(toleration.Operator) == 0 || toleration.Operator == TolerationOpEqual) && toleration.Value == taint.Value {
+	switch t.Operator {
+	case "", TolerationOpEqual:
+		return t.Value == taint.Value
+	case TolerationOpExists:
 		return true
+	default:
+		return false
 	}
-	if toleration.Operator == TolerationOpExists {
-		return true
+}
+
+// TolerationsTolerateTaint checks if taint is tolerated by any of the tolerations.
+func TolerationsTolerateTaint(tolerations []Toleration, taint *Taint) bool {
+	for i := range tolerations {
+		if tolerations[i].ToleratesTaint(taint) {
+			return true
+		}
 	}
 	return false
 }
 
-// TaintToleratedByTolerations checks if taint is tolerated by any of the tolerations.
-func TaintToleratedByTolerations(taint *Taint, tolerations []Toleration) bool {
-	tolerated := false
-	for i := range tolerations {
-		if TolerationToleratesTaint(&tolerations[i], taint) {
-			tolerated = true
-			break
+type taintsFilterFunc func(*Taint) bool
+
+// TolerationsTolerateTaintsWithFilter checks if given tolerations tolerates
+// all the interesting taints in given taint list.
+func TolerationsTolerateTaintsWithFilter(tolerations []Toleration, taints []Taint, isInterestingTaint taintsFilterFunc) bool {
+	if len(taints) == 0 {
+		return true
+	}
+
+	for i := range taints {
+		if isInterestingTaint != nil && !isInterestingTaint(&taints[i]) {
+			continue
+		}
+
+		if !TolerationsTolerateTaint(tolerations, &taints[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func DeleteTaint(taints []Taint, taintToDelete Taint) ([]Taint, bool) {
+	newTaints := []Taint{}
+	deleted := false
+	for i := range taints {
+		// if taintToRemove doesn't indicate effect, remove all the taints that have the same key
+		if (len(taintToDelete.Effect) == 0 && taintToDelete.Key == taints[i].Key) || taintToDelete.MatchTaint(taints[i]) {
```

Comment:
`(len(taintToDelete.Effect) == 0 && taintToDelete.Key == taints[i].Key)` is to support delete taints by key, which is used in kubectl taint command:
```sh
# Remove from node 'foo' all the taints with key 'dedicated'
kubectl taint nodes foo dedicated-
```
I can split this case into a new function deleteTaintsByKey() to make it less hacky: 
```go
// DeleteTaintsByKey will delete all the taints with same key to taintKey
func DeleteTaintsByKey(taints []Taint, taintKey string) ([]Taint, bool) {
	// if taint[i].Key == taintKey, delete it from the slice
}
```
DeleteTaintsByKey() is currently only called in kubectl/cmd/taint.go